server/e2e_test.go

```diff
@@ -0,0 +1,279 @@
+			h.AssertBroadcastNewPlayer(client, clients)
+		}
+	}
+	// the pool is full and the shuffle starts
+	h.AssertBroadcastPhase1Announcement(clients)
+
+	// confirm basic state on the server side
+	h.AssertPools([]poolState{
+		{
+			value:   bch1,
+			version: version,
+			players: len(clients),
+			isFull:  true,
+		},
+	})
+
+	// the shuffle succeeded, and clients leave with no blame
+	for _, c := range clients {
+		err := c.conn.Close()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// after the clients leave, the pool should be removed
+	h.AssertPools([]poolState{})
+
+	// all messages should be consumed through the assertions
+	// if anything is remaining, it is outside of specification
+	// or something unexpected happened
+	h.AssertEmptyInboxes(clients)
+}
+
+// testHarness holds the pieces required for automating a shuffle
+type testHarness struct {
+	tracker *Tracker
+	packets chan *packetInfo
+	t       *testing.T
+}
+
+type testClient struct {
+	h         *testHarness
+	conn      net.Conn
+	inbox     *testInbox
+	amount    uint64
+	version   uint64
+	session   []byte
+	playerNum uint32
+}
+
+// new TestHarness sets up the required parts for automating a shuffle
+func testNewHarness(t *testing.T, poolSize int) *testHarness {
+	// prepare shuffle environment: tracker, packet channel, connections
+	anyPort := 0
+	tracker := NewTracker(poolSize, anyPort, anyPort, anyPort, anyPort)
+	piChan := make(chan *packetInfo)
+	go startPacketInfoChan(piChan)
+
+	return &testHarness{
+		tracker: tracker,
+		packets: piChan,
+		t:       t,
+	}
+}
+
+// NewClient creates a client with an in-memory connection to server
+func (h *testHarness) NewClient() *testClient {
+	clientConn, serverConn := net.Pipe()
+
+	// handle the server side of the connection
+	go handleConnection(serverConn, h.packets, h.tracker)
+
+	// handle the client side of the connection
+	inbox := testNewInbox(clientConn)
+
+	// return a client that has done nothing but connect
+	return &testClient{
+		h:     h,
+		conn:  clientConn,
+		inbox: inbox,
+	}
+}
+
+// Register has each connection request a registration
+// https://github.com/cashshuffle/cashshuffle/wiki/CashShuffle-Server-Specification#entering-a-shuffle
+func (client *testClient) Register(amount, version uint64) {
+	client.amount = amount
+	client.version = version
+
+	testTempKey++
+	registrationMessage := &message.Signed{
+		Packet: &message.Packet{
+			FromKey: &message.VerificationKey{
+				Key: strconv.Itoa(testTempKey),
+			},
+			Registration: &message.Registration{
+				Amount:  amount,
+				Version: version,
+				Type:    message.ShuffleType_DEFAULT,
+			},
+		},
+		Signature: nil,
+	}
+
+	err := writeMessage(client.conn, []*message.Signed{registrationMessage})
+	if err != nil {
+		client.h.t.Fatal(err)
+	}
+
+	client.playerNum, client.session = client.h.AssertRegistered(client)
+}
+
+type testInbox struct {
+	packets []*packetInfo
+	mutex   sync.Mutex
+}
+
+func testNewInbox(conn net.Conn) *testInbox {
```

Comment:
This should probably be named `newTestInbox` =)

Reply:
Pending same discussion above.

Reply:
Addressed above. =)

Reply:
b17bd34 same as above